dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/ImageRegistryInterface.java

```diff
@@ -17,49 +17,415 @@
 package io.dockstore.webservice.helpers;
 
 import java.util.ArrayList;
+import java.util.Date;
+import java.util.Iterator;
 import java.util.List;
-import java.util.Map;
+import java.util.Set;
 
+import org.apache.http.client.HttpClient;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.dockstore.client.cli.nested.AbstractEntryClient;
+import io.dockstore.webservice.core.Entry;
+import io.dockstore.webservice.core.Registry;
+import io.dockstore.webservice.core.Token;
 import io.dockstore.webservice.core.Tool;
 import io.dockstore.webservice.core.Tag;
+import io.dockstore.webservice.core.ToolMode;
+import io.dockstore.webservice.core.User;
+import io.dockstore.webservice.jdbi.FileDAO;
+import io.dockstore.webservice.jdbi.TagDAO;
+import io.dockstore.webservice.jdbi.ToolDAO;
+import io.dockstore.webservice.jdbi.UserDAO;
 
 /**
- * Interface for how to grab data from a registry for docker containers.
- *
- * TODO: A pretty poor abstraction, there are quay.io data structures in here, one step at a time
- * 
+ * Abstract class for registries of docker containers.
+ **
  * @author dyuen
  */
-public interface ImageRegistryInterface {
+public abstract class ImageRegistryInterface {
+    private static final Logger LOG = LoggerFactory.getLogger(ImageRegistryInterface.class);
 
     /**
-     * Get all tags for a given tool
-     * 
-     * @return a list of tags for image that this points to
+     * Get the list of namespaces and organizations that the user is associated to on Quay.io.
+     *
+     * @return list of namespaces
      */
-    List<Tag> getTags(Tool tool);
+    public abstract List<String> getNamespaces();
 
     /**
-     * Get the list of namespaces and organizations that the user is associated to on Quay.io.
+     * Get all tags for a given tool
      *
-     * @return list of namespaces
+     * @return a list of tags for image that this points to
      */
-    List<String> getNamespaces();
+    public abstract List<Tag> getTags(Tool tool);
 
     /**
      * Get all containers from provided namespaces
-     * 
+     *
      * @param namespaces
      * @return
      */
-    List<Tool> getContainers(List<String> namespaces);
+    public abstract List<Tool> getToolsFromNamespace(List<String> namespaces);
+
+    /**
+     * Updates each tool with build/general information
+     * @param apiTools
+         */
+    public abstract void updateAPIToolsWithBuildInformation(List<Tool> apiTools);
+
+    /**
+     * Updates/Adds/Deletes tools and their associated tags
+     * @return
+     */
+    @SuppressWarnings("checkstyle:parameternumber")
+    public List<Tool> refreshTools(final Long userId, final UserDAO userDAO, final ToolDAO toolDAO,
+            final TagDAO tagDAO, final FileDAO fileDAO, final HttpClient client,
+            final Token githubToken, final Token bitbucketToken) {
+        // Get all the namespaces for the given registry
```

Comment:
This likely won't function if the user is null, so change `userId` to a primitive long (and handle a null user earlier if relevant). Put the Tokens in a Map<Enum, Token> or similar since this will likely need more if we support GitLab for example. Same with the next method